feat: Add hover info for modules and types including any docstring

plevold opened the pull request:
Fixes #208

@gnikit I decided to clone the fortls repo and have a look around. Figured I actually had a shot a making #208 work myself. Here's my suggestion. Feel free to nitpick on potential issues or anything I might have overlooked!

gnikit commented:
Thanks for the PR and the issue @plevold. I will try and have a look this week and leave a review. :+1:

gnikit commented:
The thing that you need to be extra careful is how Doxygen and modified FORD documentation can look very similar.
for example in the issue you posted you added this code

```fortran90
    !> Some documentation here?
    type :: a_t
        !> Or here?
       integer :: var    ! what if there is variable? is `!> Or here?` considered its docstrings
    end type
```

What do we expect the behaviour of the Document parser to be prefer 1 of the 2 docstrings (and if so why) or concatenate?
I remember looking into this a few years back but I don't recall what the syntax should be.

FYI
We can do **A LOT** with fortls in terms documentation. I have intentionally absteint, because users tend to get annoyed when something changes without backwards compatibility, but fortls could both auto generated docs and convert FORD to Doxygen and vice versa.

plevold commented:
> The thing that you need to be extra careful is how Doxygen and modified FORD documentation can look very similar.
> for example in the issue you posted you added this code

Agree and I see now that my initial issue description was incorrect in that regard. My changes actually doesn't do anything to the docstring parsing itself. The docstrings were already parsed correctly and added to the module and type objects. The only thing I had to add was to return the docstring in hover requests for module and type names.

One thing I noticed during my testing is that Doxygen style comments (aka. `pending_doc`) will get added to the next `Scope` even if there are other expressions in between. So for example, if somebody tries to add a docstring to an expression:

```fortran
subroutine sub_before
    integer :: i

    !> Trying to add a docstring to an expression
    i = 1
end subroutine


subroutine sub_no_doc
end subroutine
```

It will actually be added to the `sub_no_doc` subroutine below. I might try to fix this at some point, but I think that belongs to a separate PR.

gnikit commented on fortls/objects.py:
```diff
@@ -714,6 +714,14 @@ def get_type(self, no_link=False):
         return "MODULE"
 
+    def get_hover(self, long=False, include_doc=True, drop_arg=-1):
+        hover = "MODULE " + self.name
```
use fstrings and `self.get_desc`

gnikit commented on fortls/objects.py:
```diff
@@ -1318,6 +1326,19 @@ def get_actions(self, sline, eline):
             ]
         return actions
 
+    def get_hover(self, long=False, include_doc=True, drop_arg=-1):
+        keywords = ""
+        if self.abstract:
+            keywords += ", ABSTRACT"
+        if self.inherit:
+            keywords += ", EXTENDS({})".format(self.inherit)
+        hover = "TYPE{} :: {}".format(keywords, self.name)
+        if include_doc:
+            doc_str = self.get_documentation()
+            if doc_str is not None:
+                hover += "\n" + doc_str
+        return hover, long
```
use fstrings and consider using a list for keywords.
I think that explicit newline character might not be correct now that Markdown is being used for hover messages.

 Also, we need to provide some unit tests for the non-covered lines.

gnikit reviewed (changes requested):
Looks good, a couple of things are needed for this to be merged.
- rebase to master so the Markdown changes are ported
- use f-strings where possible instead of `format()` or `"str" + "val"`
- add unittests for the uncovered lines

The issue that you noted with Doxygen style comments being added to the wrong scope should also be ideally addressed since it can lead to quite a bit of a nightmare. Might be better if done in a separate PR though.

I will have some time this week after I finish the work with autocopmletions and Markdown to test your version of fortls locally and see if we need to add anything else in the hover signature and/or signature help and completions.

gnikit commented on fortls/langserver.py:
```diff
@@ -1101,6 +1101,11 @@ def create_signature_hover():
             elif var_obj.desc.endswith("STRING"):
                 hover_str = f"CHARACTER(LEN={len(var_obj.name)-2})"
                 hover_array.append(create_hover(hover_str, True))
+        else:
+            # Try and see if var_obj has any hover info. If it has, show it.
+            hover_str, highlight = var_obj.get_hover(long=True)
+            if hover_str:
+                hover_array.append(create_hover(hover_str, highlight))
```
I missed this on my review. Is this the best way for this to be done? The doc parser cannot be very robust due to their dual (FORD/Doxygen) nature so it is bound to assign incorrect documentation in multiple AST nodes. I would explicitly check of a `var_obj` is `CLASS_TYPE_ID` or `MODULE_TYPE_ID`.

Also, if it's a `SUBMODULE_TYPE_ID` how should the documentation be parsed and assigned?

plevold commented:
@gnikit thanks for the feedback and sorry about the delay! My schedule has been incredibly unpredictable the last month or so.

I think I now have resolved all the items you pointed out. As you noticed I also added #237 regarding the potential pitfalls with `pending_doc`.

gnikit commented:
There are some stuff that I need to change, style in fortran code, some changes you made that remove coverage, so if it's okay with you, I will push directly on this branch.

plevold commented:
Sure, no problem

gnikit commented:
I have no idea what's going on with the coverage reports in this PR.

I have started a new one (https://github.com/fortran-lang/fortls/pull/246) with your commit cherry-picked, but edited to undo some of the breaking changes like disabling autoupdating and `'` vs `"`. I will inspect the new coverage reports and see if any coverage lines are being removed this time around
